### PR TITLE
Add callable template support

### PR DESCRIPTION
**Pull request by mvorisek**
needed for https://github.com/phpstan/phpstan/issues/8964

based on https://github.com/phpstan/phpstan/issues/8964#issuecomment-1445362737 analysis

### CODE REVIEW

File: src/Ast/Type/CallableTypeNode.php
Comment (mvorisek) [inline src/Ast/Type/CallableTypeNode.php:25]: @ondrejmirtes is this the right OOP design or you want separate `GenericCallableTypeNode`?

`GenericTypeNode` is a litttle problematic then, as `$variances` are stored separately from `$genericTypes` there, shouldn't that be stored in another separate class?

Comment (mvorisek) [inline src/Ast/Type/CallableTypeNode.php:25]: @ondrejmirtes is this the right OOP design or you want separate `GenericCallableTypeNode`?

`GenericTypeNode` is a litttle problematic then, as `$variances` are stored separately from `$genericTypes` there, shouldn't that be stored in another separate class? /cc @jiripudil

File: doc/grammars/type.abnf
Comment (mvorisek) [inline doc/grammars/type.abnf:38]: @ondrejmirtes can I ask you for a hint how to fix the fuzzer tests? The grammar seems correct to me, and the manualy tests are passing as expected...

File: src/Parser/TypeParser.php
Comment (mvorisek) [inline src/Parser/TypeParser.php:544]: like in https://github.com/phpstan/phpdoc-parser/blob/1.22.0/src/Parser/PhpDocParser.php#L908

**Comment by mvorisek** [comment]
PR is done.

File: src/Ast/Type/CallableTypeTemplateNode.php
Comment (jiripudil) [inline src/Ast/Type/CallableTypeTemplateNode.php]: nitpick: I don't think this should be a `TypeNode`, just `Node`

File: src/Parser/TypeParser.php
Comment (jiripudil) [inline src/Parser/TypeParser.php]: Is there a benefit to this change, such as in performance? Personally, I find it more difficult to read this way. Feels like I have to think harder to figure out what's going on here. I understand that's very subjective, just wanted to raise the question :)

File: src/Parser/TypeParser.php
Comment (jiripudil) [inline src/Parser/TypeParser.php:504]: I'm wondering if this should be tied to `CallableType`? I believe the concept is more universal, and could later be reused in other places where an inline declaration of template types might be useful, such as in generic type aliases 🤔

**Review by jiripudil — COMMENTED** [review]
Hi, I took a look since this somehow ended up in my notifications and piqued my curiosity. But I'm really just passing by and only wanted to point out a few topics for discussion, so don't take my comments too seriously :)

Nice work overall!

File: src/Parser/TypeParser.php
Comment (mvorisek) [reply src/Parser/TypeParser.php]: Previously some code parts were duplicated...

Now the LOC is lower and IMO it is more readable too. In first iteration, we parse/expect no comma, in 2nd+ we parse it with delimiting comma + allow trailing comma.

File: src/Ast/Type/CallableTypeTemplateNode.php
Comment (mvorisek) [reply src/Ast/Type/CallableTypeTemplateNode.php]: Probably yes, as the class cannot be used as type itself. And it can be always, with full BC, changed back anytime ;-)

File: src/Parser/TypeParser.php
Comment (mvorisek) [reply src/Parser/TypeParser.php:504]: So you prefer to call it "template" only? @ondrejmirtes should decide, as it might be confusing with "generic" later...

Can you post any phpdoc example, outside callable, where a template can make sense?

Comment (jiripudil) [reply src/Parser/TypeParser.php:504]: Haven't thought about the naming, honestly. As for the example, as I've said, generic type aliases could be a candidate:

```php
/**
 * @phpstan-type Factory<T> = callable(): T
 */
```

I'm not planning to implement them anytime soon (if at all), but it shows that the concept of inline template type declarations could be reused. That being said, it can easily be refactored when/if the need arises

**Comment by mvorisek** [comment]
@ondrejmirtes this parser PR should be fully done, https://github.com/phpstan/phpstan/issues/8964#issuecomment-1445362737 might be easy for you, but I am not familiar with phpstan internals so I would appreciate a help with the integration into phpstan. Thank you.

**Comment by mvorisek** [comment]
@ondrejmirtes this PR was here for several weeks, may I ask you to review it?

**Comment by mvorisek** [comment]
@ondrejmirtes this PR is here for over a month, may I ask you to review it?

File: src/Printer/Printer.php
Comment (ondrejmirtes) [inline src/Printer/Printer.php]: Types need to be printed using `$this->printType` here, however here you're printing them using `TypeNode::__toString()`.

File: src/Printer/Printer.php
Comment (ondrejmirtes) [inline src/Printer/Printer.php:101]: 1) This change is not tested.
2) I believe we need to test other situations. Like when gets `CallableTypeNode` added or removed, we also need to add `<...>` and remove `<...>` around them.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
These are the two main issues I see in this PR.

Also I find the changes of the parser hard to understand, I have to inspect them in depth to understand the implications. Maybe if you could do the changes to the TypeParser in multiple smaller steps, I'd welcome them as separate PRs that I would merge before this one.

The credo to live by is: "Make the change easy, then make the easy change."

File: src/Printer/Printer.php
Comment (mvorisek) [reply src/Printer/Printer.php:101]: what test should I use as a base to start?

Comment (ondrejmirtes) [reply src/Printer/Printer.php:101]: All the tests in PrinterTest, for example:

```php
		yield [
			'/** @var string&(integer|float) */',
			'/** @var string&(int|float) */',
			new class extends AbstractNodeVisitor {

				public function enterNode(Node $node)
				{
					if ($node instanceof IdentifierTypeNode && $node->name === 'integer') {
						$node->name = 'int';
					}

					return $node;
				}

			},
		];
```

This test accepts an original PHPDoc, modifies it using the visitor, and then asserts that the printer correctly prints the modified PHPDoc.

File: src/Printer/Printer.php
Comment (mvorisek) [reply src/Printer/Printer.php]: done

File: src/Printer/Printer.php
Comment (mvorisek) [reply src/Printer/Printer.php:101]: > This change is not tested.

https://github.com/phpstan/phpdoc-parser/blob/1.24.2/src/Printer/Printer.php#L100 is not tested either, please add tests or please loosen this requirement for this contribution.

File: src/Ast/Type/CallableTypeTemplateNode.php
Comment (mvorisek) [inline src/Ast/Type/CallableTypeTemplateNode.php:13]: `TemplateTagValueNode` names it `name` of type `string`, @ondrejmirtes do you want apply any of these?

File: src/Printer/Printer.php
Comment (mvorisek) [reply src/Printer/Printer.php:101]: > I believe we need to test other situations. Like when gets CallableTypeNode added or removed, we also need to add <...> and remove <...> around them.

addressed

**Comment by mvorisek** [comment]
All feedback should be addressed.

**Comment by mvorisek** [comment]
@ondrejmirtes the feedback should be addressed, if there is anything else please let me know.

**Comment by mvorisek** [comment]
@ondrejmirtes sorry to tagging you again, can I please ask you to merge this PR if there is no feedback left?

File: src/Parser/TypeParser.php
Comment (ondrejmirtes) [inline src/Parser/TypeParser.php:515]: I don't understand why is there so much repeated code. We're already parsing template types in similar position for `@method`, there has to be a way to reuse that code: https://github.com/phpstan/phpdoc-parser/pull/160

Everything you should need is already in parseTemplateTagValue.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Parser/TypeParser.php
Comment (mvorisek) [reply src/Parser/TypeParser.php:515]: Can you please guide me how and what to reuse? The code in #160 (Parser/PhpDocParser in general) is for phpdoc instead of for type (Parser/TypeParser).

If the newly introduced Ast/Type/CallableTypeTemplateNode class is wanted (and it probably is - callable template should not consist of anything like description), then the code cannot be simplified much. As the grammar is not completely the same, basically the only parsing part to deduplicate would be `<`, `,`, `of`.

If you want, feel free to push any changes to my branch.

Comment (jiripudil) [reply src/Parser/TypeParser.php:515]: I think #160 implements the very same use case as this PR: it parses inline template tag definitions, disallowing descriptions, but accounting for everything else (bound types, defaults, ...). Perhaps it could suffice to extract the `parseTemplateTagValue` method into a `TemplateTagParser` that could be used in `TypeParser` as well?

Comment (ondrejmirtes) [reply src/Parser/TypeParser.php:515]: Yes, that logic should be extracted so that it's usable by both TypeParser and PhpDocParser.

Comment (mad-briller) [reply src/Parser/TypeParser.php:515]: @mvorisek are you going to look into this feedback or should i have a go at it?

Comment (ondrejmirtes) [reply src/Parser/TypeParser.php:515]: @mad-briller Please have a go at it :)

Comment (mvorisek) [reply src/Parser/TypeParser.php:515]: @mad-briller would you like to join the forces? Feel free to either reuse this PR as a separate one or submit a PR againts https://github.com/mvorisek/phpdoc-parser/tree/parse_generic_callable - this PR should be basically done, but the code should be deduplicated per https://github.com/phpstan/phpdoc-parser/pull/199#discussion_r1375390544 request.

**Comment by mvorisek** [comment]
Landed thru #232, thanks everyone involved!

**Pull request closed without merging**
